Lazy Tensor Implementation

bwasti opened the pull request:
Summary: A full lazy tensor implementation on the latest version of the fallback code in c10

!!! See the original PR and discussion here:
https://github.com/pytorch/pytorch/pull/25753

Test Plan: buck test mode/no-gpu //caffe2/test:lazy

Differential Revision: D18610575

ezyang commented on aten/src/ATen/core/LazyOps.cpp:
```diff
@@ -0,0 +1,28 @@
+
+bool Tensor::is_lazy() const {
+  return impl_->is_lazy();
+}
```
It is a little funny for a Tensor method to be defined all the way over here, rather than in `Tensor.cpp` where most people would expect it to be defined

ezyang replied:
Actually, it's worse than this. You define `is_lazy` in `native_functions.yaml` which means the codegen should be defining its own copy of `Tensor::is_lazy`, so now you have two definitions of `is_lazy`. Two definitions enter a linker, only one emerges... only the linker knows who...

ezyang replied:
Is there a good reason for `is_lazy` to be in `native_functions.yaml`; in particular, do we really want `Tensor.is_lazy` to exist in Python? (If you asked @zdevito, he'd say no!)

ezyang replied:
(A further aside: there isn't a good way to write functions in `native_functions.yaml` that bypass dispatch entirely and go straight to TensorImpl, as you do here. All of the functions like this are special cased in some way or another...)

bwasti replied:
>Is there a good reason for is_lazy to be in native_functions.yaml

I'd like `is_lazy` to exist in Python because we need to avoid some size checks.  E.g. in `torch/autograd/__init__.py` but further in operators like batch norm.

>A further aside: there isn't a good way to write functions in native_functions.yaml that bypass dispatch entirely

Should I just use dispatch then?

ezyang replied:
Anything is fine; this test isn't perf critical so it's not a big deal if it goes through the dispatcher.

ezyang commented on aten/src/ATen/core/DeprecatedTypeProperties.h:
```diff
@@ -34,6 +34,10 @@ class CAFFE2_API DeprecatedTypeProperties {
 
+  bool is_lazy() const {
+    return backend() == Backend::Lazy;
+  }
```
My vote is to just not add this method to the deprecated class :>

bwasti replied:
it's pretty much the cleanest way to get a check in autograd/engine.cpp working

ezyang commented on aten/src/ATen/core/LazyOps.cpp:
```diff
@@ -0,0 +1,28 @@
+}
+
+CAFFE2_API at::Tensor to_eager(at::Tensor const& self) {
+  auto lt = static_cast<LazyTensorImpl*>(self.unsafeGetTensorImpl());
```
You need to verify that `self` is a lazy tensor first before doing this cast.

ezyang commented on aten/src/ATen/core/LazyOps.cpp:
```diff
@@ -0,0 +1,28 @@
+CAFFE2_API at::Tensor to_eager(at::Tensor const& self) {
+  auto lt = static_cast<LazyTensorImpl*>(self.unsafeGetTensorImpl());
+  auto tensor = lt->to_eager();
+  TORCH_INTERNAL_ASSERT(tensor.is_variable());
```
No need, all tensors are variables now :3 :3 :3

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,57 @@
+  explicit LazyTensorImpl(TensorTypeSet ts)
+      : TensorImpl(
+            ts.add(TensorTypeId::LazyTensorId),
+            caffe2::TypeMeta::Make<float>(), // Hack to support gradients
```
HMMMMMMMMMM

ezyang replied:
So, like, if I pass a CUDA double tensor into LazyTensor, the LazyTensor will advertise itself as a CPU float tensor? That doesn't sound very kosher.

You are kind of stuck between a rock and a hard place here, because you kind of need dtype inference (along with shape inference) to propagate this information correctly. So we are probably just going to have to be edgy here. But it would be nicer if this problem could be very prominently documented and explained, starting with this very code spot.

ezyang replied:
For dtypes, though, I feel like JIT's inference should be "good enough" for most users. Is this not true?

bwasti replied:
hmm, I agree that I should advertise as CUDA double tensor. I can at least get closer with CUDA float tensor by inspecting the TensorTypeSet

I'll document it though

ezyang replied:
Still not documented?

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,57 @@
+            }
+          }
+          return ts;
+        }()) {
```
Here's a hot take: LazyTensor's TensorTypeSet should ONLY ever be LazyTensorId. None of the tensor ids of its inputs. Why? Because we never redispatch on a lazy tensor, we only ever do operations on the tensors it holds. Type set is a sort of "is a" relationship, not a "has a" relationship.

bwasti replied:
It's nice to know the underlying (lower precedent) TensorTypeSet that is being made lazy -- e.g. if its a GlowTensorId or CPUTensorId

ezyang replied:
Yeah, this is an occurrence of what I call "dispatch key" versus "type" confusion. If TensorTypeSet was called DispatchKeySet would you still be so eager to put CPUDispatchKey in it? I think not!

ezyang replied:
I still believe in this comment. I don't think the lower TensorTypeSet should be in the lazy TensorTypeSet.

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,57 @@
+#include <ATen/core/function_schema.h>
+#include <c10/core/TensorImpl.h>
+
+namespace at {
```
Can we copy paste the design doc for lazy tensor somewhere? Maybe in `LazyTensor.cpp`?

ezyang replied:
Or at the very least, a little bit of comments saying "this is a lazy tensor, this is what it represents" :)

ezyang replied:
This comment is still not done

ezyang replied:
It looks like the comments got put in `torch/csrc/LazyTensor.cpp` which is pretty far away from `aten/src/ATen/core/LazyTensor.{h,cpp}`. If you think `torch/csrc` is the right place to put the docs (I personally think the docs should be placed at the "lowest" spot where things are defined, aka `aten/src/ATen/core`) then at the very least there should be some cross-references pointing there, e.g., `See Note [Lazy tensor]`, for example

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,57 @@
+
+  explicit LazyTensorImpl(at::Tensor const& inp)
+      : LazyTensorImpl(inp.type_set()) {
+    inps_.emplace_back(inp);
```
Nit: you could just directly construct `inps_{inp}` here

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,57 @@
+    index_ = index;
+  }
+
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
```
This method... is weird. CC'ing @yf225. This should only get called if I call `y = x.detach()` on a tensor. I guess because lazy tensor currently lives after autograd, we are indeed obligated to make a shallow copy in this case (so yes, we need to implement it). But the way you are making the shallow copy here seems to imply that you will *redo* the computation in the detached node. That seems wrong: we should reuse the computation result, right?

yf225 replied:
For more context on the semantics: I believe `shallow_copy_and_detach` is only used when we want the new tensor to share the same tensor metadata (e.g. sizes / strides / storage pointer / storage_offset) with the old tensor, but with a new autograd history.

bwasti replied:
honestly I'm not really sure the purpose of this method and cargo-culted it over.

bwasti replied:
the computation will be reused, I believe, by the cache lookup

ezyang replied:
In that case, I think my comments in https://github.com/pytorch/pytorch/pull/30674/files#r356435098 apply

ezyang commented on aten/src/ATen/core/boxing/boxing.h:
```diff
@@ -48,9 +48,11 @@ using supports_boxing =
 template<class Result, class... Args>
 Result boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args, guts::enable_if_t<!supports_boxing<Result, Args...>::value, int> = 0) {
-  TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callUnboxed() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.");
+  TORCH_INTERNAL_ASSERT(false, opHandleToName(opHandle) + " Tried to call KernelFunction::callUnboxed() for a kernel that only has a boxed kernel and doesn't support calling from an unboxed API yet.");
```
You don't need to manually `+` the string here: just `opHandleToOpName(opHandle), ": Tried to call"` will be fine (`TORCH_INTERNAL_ASSERT` takes care of the cat'ing). I'm not sure if `name()` returns a `std::string` or not but if it doesn't you can make the function return a `const char*` in this case too.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3418,6 +3418,15 @@
+  variants: method
+
+- func: to_eager(Tensor self) -> Tensor
+  variants: method
```
Could we underscore these methods to make them private for now?

bwasti replied:
a trailing underscore? otherwise:

```
RuntimeError: Function '_to_lazy' starts with a single underscore and is configured to have a method on Tensor. Functions that start with  a single underscore should only be functions in the at:: namespace and not methods on Tensor!
```

ezyang replied:
Lol, I guess it would have to be a function hahaha

ezyang replied:
(Please not a trailing underscore, that means inplace!)

ezyang replied:
To clarify, yes, let's make them functions, in that case.

ezyang commented on c10/core/Backend.h:
```diff
@@ -25,7 +25,23 @@ namespace c10 {
+  MkldnnCPU,
+  Lazy,
+  NumOptions
+};
```
Backend needs to die arrrrrrr lol

ezyang commented on test/test_lazy.py:
```diff
@@ -0,0 +1,90 @@
+import time
+
+class TestLazy(TestCase):
+  def test_bench(self):
```
I'm surprised lint isn't failing you here, because we're supposed to use four spaces for Python indenting right?

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -428,11 +428,14 @@ void validate_outputs(
     if (!edge.is_valid()) continue;
 
     const auto& metadata = edge.function->input_metadata(edge.input_nr);
+    // If any input grad is lazy, we don't check shape properties
```
This comment is missing the most important part: the part where you say it's because of the HACK where we don't have accurate dtype or shape information on lazy tensors. In fact, there should probably be a pretty big Note somewhere about this state of affairs, cross-linked from here.

ezyang replied:
Still not fixed (there's a comment about this in the design doc in LazyTensor.cpp but you need a cross reference here)

ezyang replied:
Here's how I like to think about whether or not a cross reference is necessary: suppose that I'm a random developer who hasn't seen LazyTensor.cpp, and happens to be reading this code. Is it clear where I would go to find out more information about this weird line of code? If it's not obvious, I need a cross reference.

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+
+struct RegisterLazy {
+  RegisterLazy() {
+    static c10::RegistrationHandleRAII wtf = c10::Dispatcher::singleton().registerBackendFallbackKernel(
```
"wtf indeed" lol

bwasti replied:
will fix lol

ezyang commented on torch/autograd/__init__.py:
```diff
@@ -20,8 +20,9 @@
 def _make_grads(outputs, grads):
     new_grads = []
     for out, grad in zip(outputs, grads):
+        skip_check = out.is_lazy() or grad.is_lazy()
```
You could also add a cross-reference here

ezyang commented on test/test_lazy.py:
```diff
@@ -0,0 +1,90 @@
+    for _ in range(iters):
+      c = a + b
+      a = c
+    print(time.time() - t)
```
Probably don't want print statements in the test suite (I know this is intended to be a benchmarky thing, but for tests people want them to run as fast as possible.) Also... 10 iters is really short :>

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+
+namespace at {
+
+std::pair<torch::jit::Value*, std::vector<const LazyTensorImpl*>> getValue(
```
This function is called `getValue`. The `torch::jit::Value*` output makes sense. So... what's the vector of lazy tensor impl pointers then?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+    std::shared_ptr<torch::jit::Graph> g) {
+  // Convert LazyTensorImpl to node in the graph, inserting before
+  // the "dep" node.
+  auto getNode = [&g](
```
Since it does an insertion, maybe don't call this helper function `getNode`?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+      }
+    } else {
+      // TODO other IValue checks
+      if (a.val().isInt() && b.val().isInt() &&
```
Umm, why don't you just use `IValue::isSameIdentity` here?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+  const FunctionSchema& fs = op.schema();
+  if (fs.name() == "aten::to_eager") {
+    auto t = pop(*s).toTensor();
+    auto lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
```
Ah! I think your problem is that you didn't register the native implementation of `to_eager` under LazyTensorTypeId. If you did, it would take precedence, and then this if-block would be dead. Having to_eager get a specialized implementation would be clearer structure, IMO.

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,300 @@
+    drop(*s, arg_len);
+    std::vector<IValue> ret_vec;
+    for (auto i = 0; i < ret_len; ++i) {
+      auto tensor = detail::make_tensor<LazyTensorImpl>(std::move(fs), args.vec(), i);
```
This looks dangerous too... won't each output individually recompute the entire `fs(args)` operation? NB: I haven't actually read your graph processing code above, maybe you "fix it up" there. But even if you do, that's a pretty big code smell to me: rather than destroy sharing and have to reconstruct it later, LazyTensor should *preserve* sharing from the get go, so you don't have to work hard later.

bwasti replied:
(the std::move is straight up broken, gotta fix that)

what do you mean by sharing?  If there are two outputs, it'll be recomputed for each, but that can easily change going forward

ezyang replied:
Sharing = not recomputing. In another one of your responses, you mention that there's some sort of cache involved. But I don't want a cache! If I reuse a computation multiple times in my source program, with or without caching, I should reuse the computation in lazy mode.

bwasti replied:
I'll ensure the computation is computed once

I'd like to highlight that the reason for laziness is for the user to give up some more control to optimization paths, which should be free to recompute if semantically equivalent.

ezyang replied:
Sure. But we should faithfully record the original sharing/recomputation that the user wrote down, before we start busting it open :)

ezyang commented:
Overall, this looks pretty promising, but there are some parts which are pretty mysterious to me, so I'm batting this back to you for a round before I finish reviewing.

ezyang commented:
@bwasti don't forget to export your fbcode changes; I can't see them in phabricator when I'm in China ;)

zdevito reviewed (commented):
Chatted in person: there is an update to this PR with sibling handling so I will wait for that for a detailed review.

bwasti commented:
@zdevito @ezyang updated :)

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 91a1d5c:
* **3/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=91a1d5c8db2092d58439fa1a46e5b337db93f249).

### :detective: 3 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4008459) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (1/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76975154) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76975154))
<details>
<summary>
<code>Dec 18 22:45:16 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl'</code>
</summary>

```
Dec 18 22:45:15 gmake[1]: *** [caffe2/CMakeFiles/nnpack_test.dir/all] Error 2 
Dec 18 22:45:15 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:15 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:15 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:15 collect2: error: ld returned 1 exit status 
Dec 18 22:45:15 gmake[2]: *** [bin/half_float_ops_test] Error 1 
Dec 18 22:45:15 gmake[1]: *** [caffe2/CMakeFiles/half_float_ops_test.dir/all] Error 2 
Dec 18 22:45:16 [ 82%] Linking CXX executable ../bin/pattern_net_transform_test 
Dec 18 22:45:16 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:16 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:16 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:16 collect2: error: ld returned 1 exit status 
Dec 18 22:45:16 gmake[2]: *** [bin/pattern_net_transform_test] Error 1 
Dec 18 22:45:16 gmake[1]: *** [caffe2/CMakeFiles/pattern_net_transform_test.dir/all] Error 2 
Dec 18 22:45:16 gmake: *** [all] Error 2 
Dec 18 22:45:16 Traceback (most recent call last): 
Dec 18 22:45:16   File "setup.py", line 755, in <module> 
Dec 18 22:45:16     build_deps() 
Dec 18 22:45:16   File "setup.py", line 316, in build_deps 
Dec 18 22:45:16     cmake=cmake) 
Dec 18 22:45:16   File "/pytorch/tools/build_pytorch_libs.py", line 62, in build_caffe2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4008463) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (2/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76975419) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76975419))
<details>
<summary>
<code>Dec 18 22:45:20 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl'</code>
</summary>

```
Dec 18 22:45:19 [ 85%] Linking CXX executable ../bin/nnpack_test 
Dec 18 22:45:19 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:19 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:19 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:19 collect2: error: ld returned 1 exit status 
Dec 18 22:45:19 gmake[2]: *** [bin/nnpack_test] Error 1 
Dec 18 22:45:19 gmake[1]: *** [caffe2/CMakeFiles/nnpack_test.dir/all] Error 2 
Dec 18 22:45:20 [ 85%] Linking CXX executable ../bin/pattern_net_transform_test 
Dec 18 22:45:20 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:20 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:20 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:20 collect2: error: ld returned 1 exit status 
Dec 18 22:45:20 gmake[2]: *** [bin/pattern_net_transform_test] Error 1 
Dec 18 22:45:20 gmake[1]: *** [caffe2/CMakeFiles/pattern_net_transform_test.dir/all] Error 2 
Dec 18 22:45:20 gmake: *** [all] Error 2 
Dec 18 22:45:20 Traceback (most recent call last): 
Dec 18 22:45:20   File "setup.py", line 755, in <module> 
Dec 18 22:45:20     build_deps() 
Dec 18 22:45:20   File "setup.py", line 316, in build_deps 
Dec 18 22:45:20     cmake=cmake) 
Dec 18 22:45:20   File "/pytorch/tools/build_pytorch_libs.py", line 62, in build_caffe2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4008464) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (3/3)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=76975782) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=76975782))
<details>
<summary>
<code>Dec 18 22:45:25 FAILED: bin/conv_to_nnpack_transform_test</code>
</summary>

```
Dec 18 22:45:25 [2636/3002] Linking CXX executable bin/dead_code_elim_test 
Dec 18 22:45:25 FAILED: bin/dead_code_elim_test  
_code_elim_test.dir/opt/dead_code_elim_test.cc.o  -o bin/dead_code_elim_test  -Wl,-rpath,/pytorch/build/lib: lib/libgtest_main.a -Wl,--no-as-needed,/pytorch/build/lib/libtorch.so -Wl,--as-needed -Wl,--no-as-needed,/pytorch/build/lib/libtorch_cpu.so -Wl,--as-needed lib/libprotobuf.a lib/libc10.so /opt/intel/lib/libmkl_intel_lp64.a /opt/intel/lib/libmkl_gnu_thread.a /opt/intel/lib/libmkl_core.a -fopenmp -lpthread -lm /usr/lib/x86_64-linux-gnu/libdl.so lib/libmkldnn.a lib/libgtest.a -lpthread && : 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:25 collect2: error: ld returned 1 exit status 
Dec 18 22:45:25 [2637/3002] Building CXX object caffe2/CMakeFiles/transform_test.dir/core/transform_test.cc.o 
Dec 18 22:45:25 [2638/3002] Building CXX object caffe2/CMakeFiles/variant_test.dir/__/aten/src/ATen/test/variant_test.cpp.o 
Dec 18 22:45:25 [2639/3002] Linking CXX executable bin/conv_to_nnpack_transform_test 
Dec 18 22:45:25 FAILED: bin/conv_to_nnpack_transform_test  
rms/conv_to_nnpack_transform_test.cc.o  -o bin/conv_to_nnpack_transform_test  -Wl,-rpath,/pytorch/build/lib: lib/libgtest_main.a -Wl,--no-as-needed,/pytorch/build/lib/libtorch.so -Wl,--as-needed -Wl,--no-as-needed,/pytorch/build/lib/libtorch_cpu.so -Wl,--as-needed lib/libprotobuf.a lib/libc10.so /opt/intel/lib/libmkl_intel_lp64.a /opt/intel/lib/libmkl_gnu_thread.a /opt/intel/lib/libmkl_core.a -fopenmp -lpthread -lm /usr/lib/x86_64-linux-gnu/libdl.so lib/libmkldnn.a lib/libgtest.a -lpthread && : 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::~LazyTensorImpl()' 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `at::LazyTensorImpl::to_eager() const' 
Dec 18 22:45:25 /pytorch/build/lib/libtorch_cpu.so: undefined reference to `vtable for at::LazyTensorImpl' 
Dec 18 22:45:25 collect2: error: ld returned 1 exit status 
Dec 18 22:45:26 [2640/3002] Building CXX object caffe2/CMakeFiles/backend_fallback_test.dir/__/aten/src/ATen/test/backend_fallback_test.cpp.o 
Dec 18 22:45:26 [2641/3002] Building CXX object caffe2/CMakeFiles/kernel_lambda_test.dir/__/aten/src/ATen/core/boxing/kernel_lambda_test.cpp.o 
Dec 18 22:45:27 [2642/3002] Building CXX object caffe2/CMakeFiles/time_observer_test.dir/observers/time_observer_test.cc.o 
Dec 18 22:45:28 [2643/3002] Building CXX object caffe2/CMakeFiles/net_test.dir/core/net_test.cc.o 
Dec 18 22:45:29 [2644/3002] Building CXX object caffe2/CMakeFiles/elementwise_op_test.dir/operators/elementwise_op_test.cc.o 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

bwasti committed:
Lazy Tensor Implementation (#30674)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30674

A full lazy tensor implementation on the latest version of the fallback code in c10

Test Plan: buck test mode/no-gpu //caffe2/test:lazy

Differential Revision: D18610575

fbshipit-source-id: 98fe198b5ab0fb074168892cd9862fe9b597f395

ezyang commented:
Quick update: I'm back from vacation and this is on my list, but I won't get to it until after PSC.

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,111 @@
+    return impl;
+  }
+
+  inline bool isResolved() const {
```
I don't think "resolved" is standard nomenclature, so it would be good to define, semantically, what this means. (I see here that it means that the schema string is not set... but what does that *mean*?)

ezyang replied:
Based on reading below, a more standard term may be "materialized", which makes it clearer that the tensor data in question has been materialized into memory

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,111 @@
+    return schema_.value();
+  }
+
+  inline bool hasSiblings() const {
```
What are siblings?

ezyang commented on aten/src/ATen/core/LazyTensorImpl.cpp:
```diff
@@ -0,0 +1,31 @@
+
+// We notify all siblings that they don't need to update this value
+LazyTensorImpl::~LazyTensorImpl() {
+  for (auto* sibling : siblings()) {
```
This doesn't look thread safe.

ezyang replied:
I'm going to reiterate an early suggestion that I made: instead of manually threading tensors together via a sibling mechanism, it seems clearer and more direct to explicitly allocate an object representing an operation, and then have tensors reference the operation directly (and only specify what their output is). Then when you move a lazy tensor from unresolved to resolved, this proceeds by updating the *operation node*, rather than the tensor.

ezyang commented on aten/src/ATen/core/LazyTensor.h:
```diff
@@ -0,0 +1,111 @@
+
+  Tensor to_eager();
+
+  static std::function<Tensor(LazyTensorImpl*)>& getResolver();
```
What's a "resolver"?

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2671,29 +2671,17 @@
   supports_named_tensor: True
-  dispatch:
-    CPU: threshold
-    CUDA: threshold_cuda
```
What are these changes?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ * resolve the tensor has been completed.
+ *
+ * For backends that digest TorchScript IR, this design is interesting because it enables
+ * these backends to be used in PyTorch without the need for a TorchScript frontend.
```
Well, to be fair, it's not like the backend needs to build another frontend in TorchScript, right? The existing TorchScript frontend can be shared for all backends. So the value is more in eager mode

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ * For backends that digest TorchScript IR, this design is interesting because it enables
+ * these backends to be used in PyTorch without the need for a TorchScript frontend.
+ *
+ * Note: shape propagation is not supported! This means `to_eager` is required before
```
not *yet* supported, right?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ * Note: shape propagation is not supported! This means `to_eager` is required before
+ * any shape dependent code is run.  The implication is that some core modules,
+ * such as nn.BatchNorm, cannot be executed lazily because their implementations
+ * check shapes in Python.
```
I feel in a discussion about a limitation here, it's important to distinguish whether or not a limitation is fundamental, or incidental. In this case, the limitation is incidental, because in principle we could have standalone shape inference for every function in which case BatchNorm would work fine.

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ * that require state.  TODO: blacklist broken ops
+ *
+ * Note: LazyTensors cannot be used in multi-threaded environment.  This paradigm
+ * should subvert the need for parallelism at this level anyway.
```
Reapplying my earlier comment here. LazyTensors is not thread safe. But is this a *fundamental* limitation of the way it is designed, or just an accident / way to make the first implementation easier? If I think of lazy tensor as solely a system of recording an implicit graph, I don't see why it would not be thread safe; indeed, the natural implementation of this ought to be thread safe automatically!

Even when I treat lazy tensors as a form of lazy evaluation ala Haskell, it should *still* be thread safe; section 8 of https://simonmar.github.io/bib/papers/multicore-ghc.pdf gives a good example of how GHC implements lazy evaluation in a thread safe manner. Our performance requirements are much less stringent here, and a simple version of this scheme ought to work.

But I am all guessing here, because all this comment says is that "LazyTensors cannot be used in a multi-threaded environment".

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+  return true;
+}
+
+struct GraphCache {
```
So, what's the eviction policy on this cache?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+namespace at {
+
+// @returns pair of output and required stack inputs
+std::pair<torch::jit::Node*, std::vector<const LazyTensorImpl*>> getRequiredGraphValues(
```
I'm still not sure what this function does. Perhaps a more elaborated comment would be useful?

ezyang replied:
Having read the function now, the bare minimum description would be to say that this function is what implements lowering to TorchScript IR. But there are some nontrivial choices in how we go about the lowering which I feel are important to also discuss in English.

The model here seems to be that we do a BFS on the implicit graph until we reach non lazy tensors or resolved lazy tensors. Only *resolved* lazy tensors are considered as "inputs" to the graph in question; this is kind of weird, because it seems like there are two ways to end up with a resolved tensor: you called `x.to_lazy()` (this seems appropriate) or you called `y.to_eager()` and overwrote the lazy tensor with the result of the computation (this doesn't seem appropriate).

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+          auto* node = createNode(lt, dep_node);
+          dep_node->addInput(node->outputs()[lt->outputIndex()]);
+          for (const auto& input : lt->inputs()) {
+            next_frontier.emplace_back(std::make_pair(input, node));
```
You are doing the book-keeping externally without a hash map, which means you're doing a tree traversal rather than a graph traversal. This looks dangerous. What if I have:

```
x1 = x0.to_lazy()
x2 = x1 + x1
x3 = x2 + x2
# ...
```

It seems to me that you would end up with an exponentially sized graph in this case.

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ *    c = a @ b
+ *    d = c * 2
+ *   this becomes
+ *    mul[0], mm[0], 2, a, b
```
Err, you said prefix notation, is that right? Then wouldn't it be `mul[0], mm[0], a, b, 2`?

ezyang commented on torch/csrc/LazyTensor.cpp:
```diff
@@ -0,0 +1,385 @@
+ *   this becomes
+ *    mul[0], mm[0], 2, a, b
+ *   that is, all unresolved LTs in the graph are "ExprName[index]" and
+ *   resolved LTs are just their corresponding IValues
```
Use of prefix notation seems questionable to me, because it unflattens graphs into trees and can, once again, lead to exponential size representations. Wouldn't some sort of hash consing approach be better?

ezyang reviewed (changes requested):
Still seems to need work

ezyang commented:
@bwasti and I had a call where we fleshed out an alternate design for lazy tensors that helps solve some of the issues from my earlier PR.

* Instead of tracking outputs from the same operator via siblings, there will be an "Operation" node that all lazy tensors point to, to record that they're all same results of the operation
* The current mode of operation in this PR is you "resolve" a Tensor, which causes this tensor and all of its siblings to carry out their computation and then fill in the results. The proposed new mode of operation is that you have a `resolve(tensors)` which takes a list of tensors and is responsible for computing materialized results for each of the tensors in questions. To keep things simple to start, we do NOT cache the computations on the tensors.

The new API is more expressive because you can now ask for the backend accelerator to compute you the results of multiple tensors (possibly from different operations) at the same time, whereas previously you had to ask for them one at a time.

MarisaKirisame commented on test/test_lazy.py:
```diff
@@ -0,0 +1,95 @@
+        c = a
+        d = b
+
+        a = a.to_lazy()
```
Hi, I have been trying to do something pretty similar to what you are doing, however I cant get "is_lazy" to work in python side... Did you guys have any problem with that?

ezyang replied:
Hmm, that's not enough detail to really say what the problem might be.

MarisaKirisame replied:
I just assume there is some problem dealing with a function returning a single boolean... Maybe a better question is, where is is_sparse in python registered? Maybe I should register it the same way as is_sparse.

ezyang replied:
IIRC `is_sparse` is handled specially. See `torch/csrc/autograd/python_variable.cpp`
